diff --git a/CMakeLists.txt b/CMakeLists.txt
index d7d45a4..c750f0f 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -49,6 +49,7 @@ set(gRPC_BUILD_MSVC_MP_COUNT 0 CACHE STRING "The maximum number of processes for
 # Options
 option(gRPC_BUILD_TESTS "Build tests" OFF)
 option(gRPC_BUILD_CODEGEN "Build codegen" ON)
+option(gRPC_BUILD_UNUSED_LIBS "Build grpc_unsecure, grpc++_unsecure, grpc_authorization_provider" ON)
 option(gRPC_DOWNLOAD_ARCHIVES "Download archives for empty 3rd party directories" ON)
 
 set(gRPC_INSTALL_default ON)
@@ -3306,6 +3307,7 @@ endif()
 
 endif()
 
+if(gRPC_BUILD_UNUSED_LIBS)
 add_library(grpc_unsecure
   src/core/call/request_buffer.cc
   src/core/channelz/channel_trace.cc
@@ -3774,6 +3776,7 @@ if(_gRPC_PLATFORM_IOS OR _gRPC_PLATFORM_MAC)
   target_link_libraries(grpc_unsecure "-framework CoreFoundation")
 endif()
 
+endif()
 foreach(_hdr
   include/grpc/byte_buffer.h
   include/grpc/byte_buffer_reader.h
@@ -3826,7 +3829,7 @@ foreach(_hdr
 endforeach()
 
 
-if(gRPC_INSTALL)
+if(gRPC_INSTALL AND gRPC_BUILD_UNUSED_LIBS)
   install(TARGETS grpc_unsecure EXPORT gRPCTargets
     RUNTIME DESTINATION ${gRPC_INSTALL_BINDIR}
     BUNDLE DESTINATION  ${gRPC_INSTALL_BINDIR}
@@ -5257,6 +5260,7 @@ target_link_libraries(grpc++_test_util
 
 endif()
 
+if(gRPC_BUILD_UNUSED_LIBS)
 add_library(grpc++_unsecure
   src/cpp/client/call_credentials.cc
   src/cpp/client/channel_cc.cc
@@ -5350,6 +5354,7 @@ target_link_libraries(grpc++_unsecure
   ${_gRPC_PROTOBUF_LIBRARIES}
 )
 
+endif()
 foreach(_hdr
   include/grpc++/alarm.h
   include/grpc++/channel.h
@@ -5569,7 +5574,7 @@ foreach(_hdr
 endforeach()
 
 
-if(gRPC_INSTALL)
+if(gRPC_INSTALL AND gRPC_BUILD_UNUSED_LIBS)
   install(TARGETS grpc++_unsecure EXPORT gRPCTargets
     RUNTIME DESTINATION ${gRPC_INSTALL_BINDIR}
     BUNDLE DESTINATION  ${gRPC_INSTALL_BINDIR}
@@ -5579,6 +5584,7 @@ if(gRPC_INSTALL)
 endif()
 
 
+if(gRPC_BUILD_UNUSED_LIBS)
 add_library(grpc_authorization_provider
   src/core/channelz/channel_trace.cc
   src/core/channelz/channelz.cc
@@ -5908,6 +5914,7 @@ target_link_libraries(grpc_authorization_provider
   ${_gRPC_ADDRESS_SORTING_LIBRARIES}
 )
 
+endif()
 foreach(_hdr
   include/grpc/byte_buffer.h
   include/grpc/byte_buffer_reader.h
@@ -5959,7 +5966,7 @@ foreach(_hdr
 endforeach()
 
 
-if(gRPC_INSTALL)
+if(gRPC_INSTALL AND gRPC_BUILD_UNUSED_LIBS)
   install(TARGETS grpc_authorization_provider EXPORT gRPCTargets
     RUNTIME DESTINATION ${gRPC_INSTALL_BINDIR}
     BUNDLE DESTINATION  ${gRPC_INSTALL_BINDIR}
